circuits/utils/enums/circuitFunctions.enum.ts

```diff
@@ -1,18 +1,18 @@
 export enum CircuitFunctions {
```

Comment:
What do you think about something like this? We need to be able to map from ID to CircuitFunction.

```suggestion
export enum CircuitFunctions {
  EMPTY,
  STACK_YELLOW,
  STACK_RED,
  STACK_BLUE,
  TRANSFORM_YELLOW_RED,
  TRANSFORM_YELLOW_BLUE,
  TRANSFORM_RED_YELLOW,
  TRANSFORM_RED_BLUE,
  TRANSFORM_BLUE_YELLOW,
  TRANSFORM_BLUE_RED,
  TRANSFORMTWO_YELLOW_RED,
  TRANSFORMTWO_YELLOW_BLUE,
  TRANSFORMTWO_RED_YELLOW,
  TRANSFORMTWO_RED_BLUE,
  TRANSFORMTWO_BLUE_YELLOW,
  TRANSFORMTWO_BLUE_RED,
}

export function mapIndexToCircuitFunction (id: number) : CircuitFunctions {
  const circuitFunction = CircuitFunctions[id]
  if (!circuitFunction) {
    throw Error ("Invalid function ID")
  }
  else return circuitFunction as unknown as CircuitFunctions
}
```

Reply:
sorry the suggestion broke a bit, manually change it

Reply:
Having a string enum in this case allows for writing of test mocks with descriptive string arguments (https://github.com/0xMycenaean/zKubes/blob/60dbd1a22a309c7e687b568654a9da8c28371899/circuits/test/main.test.ts#L24C42-L24C42). In the upcoming PR I addressed this issue by creating a separate mapping (structure is not final). 

```
export const mutatorFunctionsMapping: MutatorFunctionNames[] = [
  { displayName: 'empty', pseudoName: CircuitFunctions.EMPTY },
  {
    displayName: 'transform: yellow -> red',
    pseudoName: CircuitFunctions.TRANSFORM_YELLOW_RED,
  },
  {
    displayName: 'transform: yellow -> blue',
    pseudoName: CircuitFunctions.TRANSFORM_YELLOW_BLUE,
  },
  {
    displayName: 'transform: red -> yellow',
    pseudoName: CircuitFunctions.TRANSFORM_RED_YELLOW,
  },
  {
    displayName: 'transform: red -> blue',
    pseudoName: CircuitFunctions.TRANSFORM_RED_BLUE,
  },
  {
    displayName: 'transform: blue -> yellow',
    pseudoName: CircuitFunctions.TRANSFORM_BLUE_YELLOW,
  },
  {
    displayName: 'transform: blue -> red',
    pseudoName: CircuitFunctions.TRANSFORM_BLUE_RED,
  },
  {
    displayName: 'stack: yellow',
    pseudoName: CircuitFunctions.STACK_YELLOW,
  },
  { displayName: 'stack: red', pseudoName: CircuitFunctions.STACK_RED },
  {
    displayName: 'stack: blue',
    pseudoName: CircuitFunctions.STACK_BLUE,
  },
  {
    displayName: 'transformTwo: yellow -> red',
    pseudoName: CircuitFunctions.TRANSFORMTWO_YELLOW_RED,
  },
  {
    displayName: 'transformTwo: yellow -> blue',
    pseudoName: CircuitFunctions.TRANSFORMTWO_YELLOW_BLUE,
  },
  {
    displayName: 'transformTwo: red -> yellow',
    pseudoName: CircuitFunctions.TRANSFORMTWO_RED_YELLOW,
  },
  {
    displayName: 'transformTwo: red -> blue',
    pseudoName: CircuitFunctions.TRANSFORMTWO_RED_BLUE,
  },
  {
    displayName: 'transformTwo: blue -> yellow',
    pseudoName: CircuitFunctions.TRANSFORMTWO_BLUE_YELLOW,
  },
  {
    displayName: 'transformTwo: blue -> red',
    pseudoName: CircuitFunctions.TRANSFORMTWO_BLUE_RED,
  },
].map((el, i) => ({ ...el, index: i }));

export function mapFunctionIndexesToNames(
  functionIndexes: number[]
): MutatorFunctionNames[] {
  return functionIndexes.map((id) => {
    const func = mutatorFunctionsMapping[id];
    if (!func) throw new Error('Missing function');
    return func;
  });
}
```

Reply:
You wouldn't need to make any changes to the tests, you can still pass in `CircuitFunctions.TRANSORM_RED` etc as an arg.
Not sure about the separate mapping but I dont have the context of the other PR

Reply:
I see. So, in the function `[gridMutator]`(https://github.com/0xMycenaean/zKubes/blob/dd65cd5b1ddc181c85e4442a65d76ac5a6e48c25/circuits/utils/gridMutator.ts#L17) having argument `args` of type `number[]` and then calling `mapFunctionIndexesToNames(args[0])`. I agree, this is a better choice, makes it easier for later development.

Reply:
Yeah, you could have it like that, or keep arg as `CircuitFunctions[]' and call `mapIdToCircuitFunction` before?
Both works, go with whatever is easier for testing

Reply:
I forgot how bad enums are in TS. That Object.values for number enum is different than the Object.values for string enum. 
I agree on the simplified accessing by index in the above example, however in the above function you are casting to `unknown` and than to `CircuitFunctions` where in fact it should stay `string'. You aren't able to access the return value (`string`) like enum - `CircuitFunctions.SOME_FUNCTION`.
I decided to go for array and then extracting a string literal type (https://github.com/0xMycenaean/zKubes/blob/56bec0672601a63c927131ca44f83d68b9cea040/circuits/types/circuitFunctions.types.ts#L43). I also added getters for safe accessing.

Reply:
I forgot how bad enums are in TS. That Object.values for number enum is different than the Object.values for string enum... 
I decided to ditch enums. I agree that we can simplify structure for better integration. I decided to go for array const assertion and then extract a string literal type. I also added getters for safe accessing. This will allow for easy use of both string literals and indexes.

Reply:
https://github.com/0xMycenaean/zKubes/blob/56bec0672601a63c927131ca44f83d68b9cea040/circuits/types/circuitFunctions.types.ts#L43

Reply:
I agree enums in TS are rubbish.

An array with getters sounds good.